transaction/std/src/memo/gift_code_cancellation.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Object for 0x0202 Gift Code Cancellation memo type
+//!
+//! This was proposed for standardization in mobilecoinfoundation/mcips/pull/32
+
+use crate::{impl_memo_type_conversions, RegisteredMemoType};
+use std::convert::TryInto;
+
+/// Memo representing the cancellation of a gift code. If a gift code is
+/// never redeemed, the sender may cancel it by sending the TxOut back
+/// to their primary address. This memo will be written to the
+/// reserved change address with 8 bytes reserved for a u64 that
+/// represents the index of the cancelled gift code TxOut
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GiftCodeCancellationMemo {
+    /// The data representing the gift code memo
+    memo_data: [u8; Self::MEMO_DATA_LEN],
+}
+
+impl RegisteredMemoType for GiftCodeCancellationMemo {
+    const MEMO_TYPE_BYTES: [u8; 2] = [0x02, 0x02];
+}
+
+impl GiftCodeCancellationMemo {
+    /// The length of the custom memo data
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Create a new gift code memo
+    pub fn new(memo_data: [u8; Self::MEMO_DATA_LEN]) -> Self {
+        GiftCodeCancellationMemo { memo_data }
+    }
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Get global TxOut index of the cancelled gift code
+    pub fn cancelled_gift_code_index(&self) -> u64 {
+        u64::from_le_bytes(self.memo_data[0..8].try_into().unwrap())
+    }
+}
+
+impl From<&[u8; 64]> for GiftCodeCancellationMemo {
+    fn from(src: &[u8; Self::MEMO_DATA_LEN]) -> Self {
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+        memo_data.copy_from_slice(src);
+        Self { memo_data }
+    }
+}
+
+impl From<GiftCodeCancellationMemo> for [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
```

Comment:
How about we make these `From<Gift*Memo>` passed by reference? It seems inconsistent to make this From impl consume its param, while the other direction (from `&[u8; 64]`) passed by reference

Reply:
Yeah actually that's what the trait specifies

Reply:
I mean `From<&GiftCodeCancellationMemo>` (with &)

Reply:
Because that would break the Into trait for RegisteredMemoType trait as designed currently by enforcing a lifetime on the input (I think). The src can be a &GiftCodeCancellationMemo even when the trait is the owned type, which I should have done!

i.e. it should be:
```rust
impl From<GiftCodeCancellationMemo> for [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
    fn from(src: GiftCodeCancellationMemo) -> [u8; GiftCodeCancellationMemo::MEMO_DATA_LEN] {
        src.memo_data
    }
}
```

Reply:
Because it breaks the Into trait. It would be better to pass from reference, but that breaks the trait somehow. To be quite honest, I'm not sure why.

Reply:
What do you mean by it breaks the trait?

Reply:
It complains about Into not being satisfied for the RegisteredMemoType trait:
`the trait bound `GiftCodeCancellationMemo: Into<[u8; 64]>` is not satisfied [E0277] the trait `Into<[u8; 64]>` is not implemented for `GiftCodeCancellationMemo``

Reply:
Yup, looks like `RegisteredMemoType` requires it: `pub trait RegisteredMemoType:
    Sized + Clone + Debug + Into<[u8; 64]> + for<'a> From<&'a [u8; 64]>`